packages/components/tree/src/tree-data-source.ts

```diff
@@ -0,0 +1,381 @@
+import { DataSource } from '@sl-design-system/data-source';
+import { type TreeNodeType } from './tree-node.js';
+
+export interface TreeDataSourceNode<T> {
+  id: unknown;
+  children?: Array<TreeDataSourceNode<T>>;
+  childrenCount?: number;
+  childrenLoading?: Promise<void>;
+  dataNode: T;
+  expandable: boolean;
+  expanded: boolean;
+  expandedIcon?: string;
+  icon?: string;
+  indeterminate?: boolean;
+  label: string;
+  lastNodeInLevel?: boolean;
+  level: number;
+  parent?: TreeDataSourceNode<T>;
+  selected?: boolean;
+  type: TreeNodeType;
+}
+
+export interface TreeDataSourceMapping<T> {
+  /**
+   * Returns the number of children. This can be used in combination with
+   * lazy loading children. This way, the tree component can show skeletons
+   * for the children while they are being loaded.
+   */
+  getChildrenCount?(item: T): number | undefined;
+
+  /** Optional method for returning a custom icon for a tree node. */
+  getIcon?(item: T, expanded: boolean): string;
+
+  /** Used to identify a tree node. */
+  getId(item: T): unknown;
+
+  /**
+   * Returns a string that is used as the label for the tree node.
+   * If you want to customize how the tree node is rendered, you can
+   * provide your own `TreeItemRenderer` function to the tree component.
+   */
+  getLabel(item: T): string;
+
+  /** Returns whether the given node is expandable. */
+  isExpandable(item: T): boolean;
+
+  /**
+   * Returns whether the given node is expanded. This is only used for the initial
+   * expanded state of the node. If you want to expand/collapse a node programmatically,
+   * use the `expand` and `collapse` methods on the tree model.
+   */
+  isExpanded?(item: T): boolean;
+
+  /**
+   * Returns whether the given node is selected. This is only used for the initial
+   * selected state of the node. If you want to select/deselect a node programmatically,
+   * use the `select` and `deselect` methods on the tree model.
+   */
+  isSelected?(item: T): boolean;
+}
+
+export interface TreeDataSourceOptions<T> {
+  /** Provide this method to lazy load child nodes when a parent node is expanded. */
+  loadChildren?(node: TreeDataSourceNode<T>): Promise<Array<TreeDataSourceNode<T>>>;
+
+  /** Enables single or multiple selection of tree nodes. */
+  selects?: 'single' | 'multiple';
+}
+
+/**
+ * Abstract class used to provide a common interface for tree data.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export abstract class TreeDataSource<T = any> extends DataSource<T, TreeDataSourceNode<T>> {
+  #loadChildren?: TreeDataSourceOptions<T>['loadChildren'];
```

Comment:
Can you please add spacing between variables and maybe add descriptions as well?